tcptracer-bpf.c

```diff
@@ -0,0 +1,927 @@
+
+#define GUESS_SADDR      0
+#define GUESS_DADDR      1
+#define GUESS_FAMILY     2
+#define GUESS_SPORT      3
+#define GUESS_DPORT      4
+#define GUESS_NETNS      5
+#define GUESS_DADDR_IPV6 6
+
+struct tcp_ipv4_event_t {
+	/* timestamp must be the first field, the sorting depends on it */
+	u64 timestamp;
+	u64 cpu;
+	u32 type;
+	u32 pid;
+	char comm[TASK_COMM_LEN];
+	u32 saddr;
+	u32 daddr;
+	u16 sport;
+	u16 dport;
+	u32 netns;
+};
+
+struct tcp_ipv6_event_t {
+	/* timestamp must be the first field, the sorting depends on it */
+	u64 timestamp;
+	u64 cpu;
+	u32 type;
+	u32 pid;
+	char comm[TASK_COMM_LEN];
+	/* Using the type unsigned __int128 generates an error in the ebpf verifier */
+	u64 saddr_h;
+	u64 saddr_l;
+	u64 daddr_h;
+	u64 daddr_l;
+	u16 sport;
+	u16 dport;
+	u32 netns;
+};
+
+// tcp_set_state doesn't run in the context of the process that initiated the
+// connection so we need to store a map TUPLE -> PID to send the right PID on
+// the event
+struct ipv4_tuple_t {
+	u32 saddr;
+	u32 daddr;
+	u16 sport;
+	u16 dport;
+	u32 netns;
+};
+
+struct ipv6_tuple_t {
+	/* Using the type unsigned __int128 generates an error in the ebpf verifier */
+	u64 saddr_h;
+	u64 saddr_l;
+	u64 daddr_h;
+	u64 daddr_l;
+	u16 sport;
+	u16 dport;
+	u32 netns;
+};
+
+struct pid_comm {
+	u64 pid;
+	char comm[TASK_COMM_LEN];
+};
+
+/* This is a key/value store with the keys being the cpu number
+ * and the values being a perf file descriptor.
+ */
+struct bpf_map_def SEC("maps/tcp_event_ipv4") tcp_event_ipv4 = {
+	.type = BPF_MAP_TYPE_PERF_EVENT_ARRAY,
+	.key_size = sizeof(int),
+	.value_size = sizeof(__u32),
+	.max_entries = 1024,
+};
+
+/* This is a key/value store with the keys being the cpu number
+ * and the values being a perf file descriptor.
+ */
+struct bpf_map_def SEC("maps/tcp_event_ipv6") tcp_event_ipv6 = {
+	.type = BPF_MAP_TYPE_PERF_EVENT_ARRAY,
+	.key_size = sizeof(int),
+	.value_size = sizeof(__u32),
+	.max_entries = 1024,
+};
+
+/* These maps are used to match the kprobe & kretprobe of connect */
+struct bpf_map_def SEC("maps/connectsock_ipv4") connectsock_ipv4 = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(__u64),
+	.value_size = sizeof(void *),
+	.max_entries = 1024,
+};
+
+struct bpf_map_def SEC("maps/connectsock_ipv6") connectsock_ipv6 = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(__u64),
+	.value_size = sizeof(void *),
+	.max_entries = 1024,
+};
+
+struct bpf_map_def SEC("maps/tuplepid_ipv4") tuplepid_ipv4 = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct ipv4_tuple_t),
+	.value_size = sizeof(struct pid_comm),
+	.max_entries = 1024,
+};
+
+struct bpf_map_def SEC("maps/tuplepid_ipv6") tuplepid_ipv6 = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct ipv6_tuple_t),
+	.value_size = sizeof(struct pid_comm),
+	.max_entries = 1024,
+};
+
+/* http://stackoverflow.com/questions/1001307/detecting-endianness-programmatically-in-a-c-program */
+static inline bool is_big_endian(void)
```

Comment:
Can't this be figured out by a kernel symbol? (probably not worth the pain if it's complicated, but just in case)